### ISSUE TITLE
Manual assert_eq! (enhance `if_then_panic`)

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#7716]
We just [added](https://github.com/rust-lang/rust-clippy/pull/7669) `if_then_panic`, but then I realized there's a possible enhancement for `assert_eq!`/`assert_ne!`:

```rust
// before
if a != b {
    panic!("hi");
}
// after
assert_eq!(a, b, "hi");
```

I think a rename to `manual_assert` would be best. But we could also keep the current name (it still kinda makes sense) or create a new lint for this case.

**Comment by xFrednet** [comment]
Nice suggestion, and renaming the lint to `manual_assert` sounds like a good idea. We should try to do this, before the `if_then_panic` lint hit's stable, IMO :upside_down_face:

**Comment by Labelray** [comment]
I am interested in implementing the enhancement. And I think it is better to use just one lint `manual_assert` for current `if_then_panic` and this new case, not new lint for this new case.

@rustbot claim

**Comment by xFrednet** [comment]
Hey @Labelray, if you're working on this issue, you might also want to take a look at #7718. @mikerite suggested in the issue to move this lint to `pedantic` which seems reasonable to me. That might be an easy change to do along the way :upside_down_face:.

**Comment by camsteffen** [comment]
Another thing I thought of - lint if-else-panic and **don't** lint if-panic-else-panic.

**Comment by jplatte** [comment]
Relatedly, `if !something { panic!(); }` should suggest `assert!(something);` instead of `assert!(!!something);`. (saw this in https://github.com/ruma/ruma/runs/3759040805?check_suite_focus=true#step:5:399, I'm *guessing* that the double negation would be happening without the binop which is also not being treated correctly (#7731).

**Comment by spearman** [comment]
An issue it seems with the current `manual_assert` is that it suggests using assert for types that are not Debug.

**Comment by abendrothj** [comment]
Hi, is this still needed? I'd be happy to take a look.
